cli/cmd/cmds/devx.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	// start executing the command
+	execCmd := exec.Command(executorCmd, formatArgs(executorArgs, cmd.content)...)
+
+	stdout, err := execCmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := execCmd.Start(); err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(stdout)
+	for scanner.Scan() {
+		fmt.Println(scanner.Text())
+	}
+
+	if err := scanner.Err(); err != nil {
+		fmt.Println("error reading output:", err)
+	}
+
+	if err := execCmd.Wait(); err != nil {
+		fmt.Println("error waiting for command:", err)
+	}
+
+	return nil
+}
+
+type commandGroup struct {
+	name     string
+	commands []command
+}
+
+func (cg *commandGroup) GetId() string {
+	var result []rune
+
+	for _, char := range cg.name {
+		if unicode.IsLetter(char) || unicode.IsDigit(char) {
+			result = append(result, unicode.ToLower(char))
+		} else if unicode.IsSpace(char) {
+			result = append(result, '-')
+		}
+	}
+
+	joined := string(result)
+
+	re := regexp.MustCompile(`-+`)
+	joined = re.ReplaceAllString(joined, "-")
+
+	return strings.Trim(joined, "-")
+}
+
+func extractCommandGroups(data []byte) ([]commandGroup, error) {
+	md := goldmark.New()
+	reader := text.NewReader(data)
+	doc := md.Parser().Parse(reader)
+
+	// store the command groups and commands
+	groups := []commandGroup{}
+	var currentPlatform *string
+
+	// walk through the ast nodes
+	ast.Walk(doc, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
+		// look up for headers
+		if heading, ok := n.(*ast.Heading); ok && entering {
+			if heading.Level == 3 {
+				currentPlatform = nil
+				commandName := string(heading.Text(data))
+
+				groups = append(groups, commandGroup{
+					name:     commandName,
+					commands: []command{},
+				})
+			}
+
+			/* if heading.Level == 4 && len(groups) > 0 {
+				platform := string(heading.Text(data))
+				currentPlatform = &platform
+			} */
+		}
+
+		// look up for code blocks
+		if block, ok := n.(*ast.FencedCodeBlock); ok && entering && len(groups) > 0 {
+			i := len(groups) - 1
+			lang := string(block.Language(data))
+
+			var buf bytes.Buffer
+			for i := 0; i < block.Lines().Len(); i++ {
+				line := block.Lines().At(i)
+				buf.Write(line.Value(data))
+			}
+
+			groups[i].commands = append(groups[i].commands, command{
+				content:  buf.String(),
+				lang:     &lang,
+				platform: currentPlatform,
+			})
+		}
+
+		return ast.WalkContinue, nil
+	})
+
+	if len(groups) == 0 {
+		return nil, errors.New("no command groups found in the markdown")
+	}
+
+	return groups, nil
+}
+
+func processCmd(list []commandGroup, cmd string) error {
+	var foundCmd *command
+	for _, v := range list {
+		if v.GetId() == cmd {
+			// TODO: should get the fisrt (most specified) command corresponding to the current host platform
+			foundCmd = &v.commands[0]
+		}
+	}
+
+	if foundCmd == nil {
+		return fmt.Errorf("command not found")
+	}
+
+	return foundCmd.Exec()
+}
+
+func getLangExecutor(lang *string) (string, []string) {
```

Comment:
You might consider adopting a pattern like is [shown here](https://github.com/input-output-hk/catalyst-forge/blob/master/lib/project/secrets/client.go). In that example, we are mapping provider names (like `aws`) to functions that return structs that implement the `SecretProvider` interface. 

You can imitate this with something like a `LanguageExecutor` interface and then a map which maps things like `sh` to the `ShellLanguageExecutor`, for example. 

This pattern just makes adding additional languages trivial and standardizes things a bit.